sky/spot/controller.py

```diff
@@ -20,55 +20,65 @@
-    return task, task.name
+def _get_dag_and_name(dag_yaml: str) -> Tuple['sky.Dag', str]:
+    dag = dag_utils.load_chain_dag_from_yaml(dag_yaml)
+    dag_name = dag.name
+    assert dag_name is not None, dag
+    return dag, dag_name
 
 
 class SpotController:
     """Each spot controller manages the life cycle of one spot cluster (job)."""
 
-    def __init__(self, job_id: int, task_yaml: str,
+    def __init__(self, job_id: int, dag_yaml: str,
                  retry_until_up: bool) -> None:
         self._job_id = job_id
-        self._task, self._task_name = _get_task_and_name(task_yaml)
+        self._dag, self._dag_name = _get_dag_and_name(dag_yaml)
+        logger.info(self._dag)
 
         self._retry_until_up = retry_until_up
         # TODO(zhwu): this assumes the specific backend.
         self._backend = cloud_vm_ray_backend.CloudVmRayBackend()
 
+        # pylint: disable=line-too-long
         # Add a unique identifier to the task environment variables, so that
         # the user can have the same id for multiple recoveries.
-        #   Example value: sky-2022-10-04-22-46-52-467694_id-17
-        task_envs = self._task.envs or {}
-        job_id_env_var = common_utils.get_global_job_id(
-            self._backend.run_timestamp, 'spot', str(self._job_id))
-        task_envs[constants.JOB_ID_ENV_VAR] = job_id_env_var
-        self._task.update_envs(task_envs)
-
-        spot_state.set_submitted(
-            self._job_id,
-            self._task_name,
-            self._backend.run_timestamp,
-            resources_str=backend_utils.get_task_resources_str(self._task))
-        logger.info(f'Submitted spot job; SKYPILOT_JOB_ID: {job_id_env_var}')
-        self._cluster_name = spot_utils.generate_spot_cluster_name(
-            self._task_name, self._job_id)
-        self._strategy_executor = recovery_strategy.StrategyExecutor.make(
-            self._cluster_name, self._backend, self._task, retry_until_up)
-
-    def _run(self):
+        #   Example value: sky-2022-10-04-22-46-52-467694_spot-_id-17-1
+        #   Example value (multi-job): sky-2022-10-04-22-46-52-467694_spot_id-17-2
+        job_id_env_vars = []
+        for i in range(len(self._dag.tasks)):
+            job_id_env_var = common_utils.get_global_job_id(
+                self._backend.run_timestamp,
+                'spot',
+                str(self._job_id),
+                sub_id=i)
+            job_id_env_vars.append(job_id_env_var)
+
+        for i, task in enumerate(self._dag.tasks):
+            task_envs = task.envs or {}
+            task_envs[constants.JOB_ID_ENV_VAR_DEPRECATED] = job_id_env_vars[i]
+            task_envs[constants.JOB_ID_ENV_VAR] = job_id_env_vars[i]
+            task_envs[constants.JOB_ID_LIST_ENV_VAR] = '\n'.join(
+                job_id_env_vars)
+            task.update_envs(task_envs)
+
+    def _run_one_task(self, task_id: int, task: 'sky.Task') -> bool:
         """Busy loop monitoring spot cluster status and handling recovery.
```

Comment:
I don't think we want to clean up the bucket here as well (may need discussion), because the tasks may share the buckets, and cleaning them up after the current task succeeded may cause problems for the following tasks.